firedrake/functionspaceimpl.py

```diff
@@ -860,17 +860,35 @@ class RestrictedFunctionSpace(FunctionSpace):
     output of the solver.
 
     :arg function_space: The :class:`FunctionSpace` to restrict.
+    :kwarg boundary_set: An iterable of DirichletBCs or a set of subdomains on
+        which a DirichletBC will be applied.
     :kwarg name: An optional name for this :class:`RestrictedFunctionSpace`,
         useful for later identification.
-    :kwarg boundary_set: A set of subdomains on which a DirichletBC will be
-        applied.
 
     Notes
     -----
     If using this class to solve or similar, a list of DirichletBCs will still
     need to be specified on this space and passed into the function.
     """
-    def __init__(self, function_space, name=None, boundary_set=frozenset()):
+    def __new__(cls, *args, boundary_set=frozenset(), name=None, **kwargs):
+        if cls is not RestrictedFunctionSpace:
+            return super().__new__(cls)
+        function_space = args[0]
+        if len(function_space) > 1:
+            return MixedFunctionSpace([cls(Vsub, boundary_set=boundary_set)
+                                       for Vsub in function_space], name=name)
+        return super().__new__(cls)
+
+    def __init__(self, function_space, boundary_set=frozenset(), name=None):
+        if all(hasattr(bc, "sub_domain") for bc in boundary_set):
```

Comment:
@ksagiyam I'm not sure if importing DirichletBC on this file is a good idea

Reply:
`DirichletBC`s are big objects, so it would be nicer if `impl.RestrictedFunctionSpace` does not need to handle them, I think.
Though I said it would be good to have `RestrictedFunctionSpace(mixed_space, ...)` interface, I was not thinking too hard. Given that `RestrictedFunctionSpace`s are almost the same as regular `FunctionSpace`s (they basically only have different DoF orderings), they should just be viewed as sisters of regular `FunctionSpace`s, and we should probably just allow for `MixedFunctionSpace([V_regular, V_restrict, V_regular, ...])` interface.
So, instead of changing the `impl.RestrictedFunctionSpace` constructor, can we simply convert the `MixedFunctionSpace` object all composed of regular function spaces to one composed of regular and restricted spaces as necessary inspecting `bc.function_space` in `NonlinearVariationalProblem` around here https://github.com/firedrakeproject/firedrake/blob/ed826a575a9314b45e8e396d1a2e9da8687f524d/firedrake/variational_solver.py#L91C45-L91C46?

Reply:
I'm not changing `__init__`, but just `__new__`. I think I achieved what you just described but through `RestrictedFunctionSpace(MixedFunctionSpace, boundary_set=bcs)`. The bcs do not live within the `RestrictedFunctionSpace`, it is just easier to pass a single list as opposed to having to create separate `boundary_sets` for each mixed component.

Reply:
I suggest that we do not allow for `RestrictedFunctionSpace(MixedFunctionSpace, ...)` interface for the reason mentioned in the above. Passing `bcs` does not sound right either, no matter how we use it, as `impl.RestrictedFunctionSpace` is topological, while `bcs` contain geometric information.

Reply:
I think we could follow the same approach we have for `FunctionSpace`, and implement a user-facing function  `RestrictedFunctionSpace` in functionspace.py that parses the bcs into a `boundary_set` for each component, and constructs the MixedSpace if a MixedSpace is provided.

Reply:
So why do you think we should allow for `RestrictedFunctionSpace(MixedFunctionSpace, ...)` interface?

Reply:
It would be the most natural and intuitive way to restrict a space, and would just work in general.